File: Fibonacci Heaps/FibonacciExtractMin.java
// File: FibonacciExtractMin.java
// Program: Illustrate extracting the minimum key from a Fibonacci Heap

class Node {
    int key;
    Node left, right;

    Node(int key) {
        this.key = key;
        left = right = this;
    }
}

public class FibonacciExtractMin {

    static Node min = null;

    static void insert(int key)
    {
        Node n = new Node(key);

        if (min == null)
            min = n;
        else
        {
            n.left = min;
            n.right = min.right;
            min.right.left = n;
            min.right = n;

            if (n.key < min.key)
                min = n;
        }
    }

    static void extractMin()
    {
        if (min == null)
        {
            System.out.println("Heap is empty");
            return;
        }

        Node temp = min;
        System.out.println("Extracted minimum key: " + temp.key);

        if (temp.right == temp)
            min = null;
        else
        {
            temp.left.right = temp.right;
            temp.right.left = temp.left;
            min = temp.right;
        }
    }

    static void display()
    {
        if (min == null)
        {
            System.out.println("Heap is empty");
            return;
        }

        Node temp = min;

        System.out.print("Heap nodes: ");
        do {
            System.out.print(temp.key + " ");
            temp = temp.right;
        } while (temp != min);

        System.out.println();
    }

    public static void main(String[] args)
    {
        insert(10);
        insert(3);
        insert(15);
        insert(6);

        display();

        extractMin();

        display();
    }
}

File: Fibonacci Heaps/fibonacci_extract_min.c
/* File: fibonacci_extract_min.c
   Program: Illustrate extracting the minimum key from a Fibonacci Heap
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    int key;
    struct node *left;
    struct node *right;
} Node;

Node *min = NULL;

Node* createNode(int key)
{
    Node *n = (Node*)malloc(sizeof(Node));
    n->key = key;
    n->left = n;
    n->right = n;
    return n;
}

void insert(int key)
{
    Node *n = createNode(key);

    if (min == NULL)
        min = n;
    else
    {
        n->left = min;
        n->right = min->right;
        min->right->left = n;
        min->right = n;

        if (n->key < min->key)
            min = n;
    }
}

void extractMin()
{
    if (min == NULL)
    {
        printf("Heap is empty\n");
        return;
    }

    Node *temp = min;
    printf("Extracted minimum key: %d\n", temp->key);

    if (temp->right == temp)
        min = NULL;
    else
    {
        temp->left->right = temp->right;
        temp->right->left = temp->left;
        min = temp->right;
    }

    free(temp);
}

void display()
{
    if (min == NULL)
    {
        printf("Heap is empty\n");
        return;
    }

    Node *temp = min;

    printf("Heap nodes: ");
    do
    {
        printf("%d ", temp->key);
        temp = temp->right;
    }
    while (temp != min);

    printf("\n");
}

int main()
{
    insert(10);
    insert(3);
    insert(15);
    insert(6);

    display();

    extractMin();

    display();

    return 0;
}
